File: apps/mobile/src/services/tokenStorage.ts
import * as SecureStore from 'expo-secure-store';
import { Platform } from 'react-native';

const ACCESS_TOKEN_KEY = 'shikkhok_access_token';
const REFRESH_TOKEN_KEY = 'shikkhok_refresh_token';

let memoryAccessToken: string | null = null;
let memoryRefreshToken: string | null = null;

export const tokenStorage = {
  getAccessToken: async (): Promise<string | null> => {
    if (Platform.OS === 'web') {
      return memoryAccessToken;
    }
    try {
      return await SecureStore.getItemAsync(ACCESS_TOKEN_KEY);
    } catch {
      return memoryAccessToken;
    }
  },

  setAccessToken: async (token: string): Promise<void> => {
    memoryAccessToken = token;
    if (Platform.OS !== 'web') {
      try {
        await SecureStore.setItemAsync(ACCESS_TOKEN_KEY, token);
      } catch {
        // Fallback to memory
      }
    }
  },

  getRefreshToken: async (): Promise<string | null> => {
    if (Platform.OS === 'web') {
      return memoryRefreshToken;
    }
    try {
      return await SecureStore.getItemAsync(REFRESH_TOKEN_KEY);
    } catch {
      return memoryRefreshToken;
    }
  },

  setRefreshToken: async (token: string): Promise<void> => {
    memoryRefreshToken = token;
    if (Platform.OS !== 'web') {
      try {
        await SecureStore.setItemAsync(REFRESH_TOKEN_KEY, token);
      } catch {
        // Fallback to memory
      }
    }
  },

  clearTokens: async (): Promise<void> => {
    memoryAccessToken = null;
    memoryRefreshToken = null;
    if (Platform.OS !== 'web') {
      try {
        await SecureStore.deleteItemAsync(ACCESS_TOKEN_KEY);
        await SecureStore.deleteItemAsync(REFRESH_TOKEN_KEY);
      } catch {
        // Ignore delete errors
      }
    }
  },
};

File: apps/mobile/lib/core/storage/token_storage.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

class TokenStorage {
  static const _storage = FlutterSecureStorage(
    aOptions: AndroidOptions(
      encryptedSharedPreferences: true,
    ),
    iOptions: IOSOptions(
      accessibility: KeychainAccessibility.first_unlock,
    ),
  );

  static const _accessTokenKey = 'shikkhok_access_token';
  static const _refreshTokenKey = 'shikkhok_refresh_token';
  static const _deviceIdKey = 'shikkhok_device_id';
  static const _userRoleKey = 'shikkhok_user_role';
  static const _userIdKey = 'shikkhok_user_id';

  static Future<void> saveTokens({
    required String accessToken,
    required String refreshToken,
  }) async {
    await _storage.write(key: _accessTokenKey, value: accessToken);
    await _storage.write(key: _refreshTokenKey, value: refreshToken);
  }

  static Future<void> setAccessToken(String token) async {
    await _storage.write(key: _accessTokenKey, value: token);
  }

  static Future<String?> getAccessToken() async {
    return await _storage.read(key: _accessTokenKey);
  }

  static Future<void> setRefreshToken(String token) async {
    await _storage.write(key: _refreshTokenKey, value: token);
  }

  static Future<String?> getRefreshToken() async {
    return await _storage.read(key: _refreshTokenKey);
  }

  static Future<bool> hasAccessToken() async {
    final token = await getAccessToken();
    return token != null && token.isNotEmpty;
  }

  static Future<void> saveUserMetadata({
    required String userId,
    required String userRole,
  }) async {
    await _storage.write(key: _userIdKey, value: userId);
    await _storage.write(key: _userRoleKey, value: userRole);
  }

  static Future<String?> getUserId() async {
    return await _storage.read(key: _userIdKey);
  }

  static Future<String?> getUserRole() async {
    return await _storage.read(key: _userRoleKey);
  }

  static Future<String> getOrCreateDeviceId() async {
    var deviceId = await _storage.read(key: _deviceIdKey);
    if (deviceId == null || deviceId.isEmpty) {
      deviceId = 'flutter-dev-${DateTime.now().millisecondsSinceEpoch}';
      await _storage.write(key: _deviceIdKey, value: deviceId);
    }
    return deviceId;
  }

  static Future<void> clearTokens() async {
    await _storage.delete(key: _accessTokenKey);
    await _storage.delete(key: _refreshTokenKey);
    await _storage.delete(key: _userIdKey);
    await _storage.delete(key: _userRoleKey);
  }
}
